File: shared/src/androidMain/kotlin/com/commuto/interfacemobile/kmService/kmTypes/SymmetricKey.kt
package com.commuto.interfacemobile.kmService.kmTypes

import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * The SymmetricKey class is a wrapper around an AES-256 SecretKey, with methods for encrypting and decrypting data
 * using the key.
 *
 * @property key: the SecretKey object around which this class is wrapped.
 * @property keyBytes: a ByteArray representation of the wrapped SecretKey.
 */
class SymmetricKey {
    val key: SecretKey
    val keyBytes: ByteArray

    constructor (key: SecretKey) {
        this.key = key
        this.keyBytes = key.encoded
    }

    constructor(key: ByteArray) {
        this.keyBytes = key
        this.key = SecretKeySpec(key, "AES")
    }

    /**
     * Encrypt data with this key using CBC with PKCS5 padding.
     *
     * @param data: the data to be encrypted
     * @returns SymmetricallyEncryptedData: the encrypted data along with the initialization vector used.
     */
    fun encrypt(data: ByteArray): SymmetricallyEncryptedData {
        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.ENCRYPT_MODE, this.key)
        val iv = cipher.iv
        val encryptedBytes = cipher.doFinal(data)
        return SymmetricallyEncryptedData(encryptedBytes, iv)
    }

    /**
     * Decrypt data encrypted using CBC with PKCS5 padding with this key.
     *
     * @param data: the SymmetricallyEncryptedData object containing both the encrypted data and the initialization
     * vector used.
     * @returns ByteArray: the decrypted data
     */
    fun decrypt(data: SymmetricallyEncryptedData): ByteArray {
        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.DECRYPT_MODE, this.key, IvParameterSpec(data.initializationVector))
        return cipher.doFinal(data.encryptedData)
    }

}

/**
 * The SymmetricallyEncryptedData class contains symmetrically encrypted data and the initialization vector used to
 * encrypt it.
 *
 * @property encryptedData: symmetrically encrypted data
 * @property iv: the initialization vector used to encrypt encryptedData
 */
class SymmetricallyEncryptedData(data: ByteArray, iv: ByteArray) {
    val encryptedData = data
    val initializationVector = iv
}

/**
 * Returns a new SymmetricKey object wrapped around an AES-256 SecretKey.
 */
fun newSymmetricKey(): SymmetricKey {
    val keyGenerator = KeyGenerator.getInstance("AES")
    keyGenerator.init(256)
    return SymmetricKey(keyGenerator.generateKey())
}

File: iosApp/iosApp/kmService/kmTypes/SymmetricKey.swift
//
//  SymmetricKey.swift
//  iosApp
//
//  Created by jimmty on 12/2/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import CryptoSwift
import Foundation

/**
 * The SymmetricKey struct contains an AES-256 key, with methods for encrypting and decrypting data using the key.
 *
 * - Parameter keyBytes: the AES-256 key in byte array format
 * - Parameter keyData: the AES-256 key in Data format
 */
struct SymmetricKey {
    let keyBytes: [UInt8]
    let keyData: Data
    
    init(key: [UInt8]) {
        self.keyBytes = key
        self.keyData = Data(keyBytes)
    }
    
    init(key: Data) {
        self.keyData = key
        self.keyBytes = key.bytes
    }

    /**
     * Encrypt data with this key using CBC with PKCS5 padding.
     *
     * - Parameter data: the data to be encrypted
     * - Returns SymmetricallyEncryptedData: the encrypted data along with the initialization vector used.
     */
    func encrypt(data: Data) throws -> SymmetricallyEncryptedData {
        let iv: [UInt8] = AES.randomIV(AES.blockSize)
        let aes = try AES(key: keyBytes, blockMode: CBC(iv: iv), padding: .pkcs5)
        let encryptedBytes = try aes.encrypt(data.bytes)
        return SymmetricallyEncryptedData(data: encryptedBytes, iv: iv)
    }
    
    /**
     * Decrypt data encrypted using CBC with PKCS5 padding with this key.
     *
     * - Parameter data: the SymmetricallyEncryptedData object containing both the encrypted data and the initialization
     * vector used.
     * - Returns Data: the decrypted data
     */
    func decrypt(data: SymmetricallyEncryptedData) throws -> Data {
        let aes = try AES(key: keyBytes, blockMode: CBC(iv: data.initializationVectorBytes), padding: .pkcs5)
        return Data(try aes.decrypt(data.encryptedDataBytes))
    }
}

/**
 * The SymmetricallyEncryptedData struct contains symmetrically encrypted data and the initialization vector used to encrypt it.
 *
 * - Parameter encryptedData: symmetrically encrypted data
 * - Parameter iv: the initialization vector used to encrypt encryptedData
 */
struct SymmetricallyEncryptedData {
    let encryptedDataBytes: [UInt8]
    let encryptedData: Data
    
    let initializationVectorBytes: [UInt8]
    let initializationVectorData: Data
    
    init(data: [UInt8], iv: [UInt8]) {
        self.encryptedDataBytes = data
        self.initializationVectorBytes = iv
        
        self.encryptedData = Data(encryptedDataBytes)
        self.initializationVectorData = Data(initializationVectorBytes)
    }
    
    init(data: Data, iv: Data) {
        self.encryptedDataBytes = data.bytes
        self.initializationVectorBytes = iv.bytes
        
        self.encryptedData = data
        self.initializationVectorData = iv
    }
}

/**
 Thrown when newSymmetricKey() encounters an error generating random bytes
 */
enum SymmetricKeyError: Error {
    case creationError(message: String)
}

/**
 * Returns a new SymmetricKey object wrapped around an AES-256 key.
 */
func newSymmetricKey() throws -> SymmetricKey {
    var keyBytes = [UInt8](repeating: 0, count: 32)
    let status = SecRandomCopyBytes(kSecRandomDefault, keyBytes.count, &keyBytes)
    if status != errSecSuccess {
        throw SymmetricKeyError.creationError(message: "Unexpected error generating random bytes")
    }
    return SymmetricKey(key: keyBytes)
}
